### PR TITLE
[BE] create register component

### PR DESCRIPTION
**Pull request by miguel1808**
Introduced new user and password hash functions using bcrypt

UserRepository interface was established to insert new users into the database

Use case for user verification by email to validate that multiple accounts are not created with the same email

The registerUserHandler handler was developed to send HTTP registration requests,

### CODE REVIEW

**Comment by luzelenatt** [comment]
@AngelGio

File: backend/internal/register/application/registerUseCase.go [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+import (
+	"errors"
+
+	"github.com/unmsmfisi-socialapplication/social_app/internal/register/domain"
+)
+
+var (
+	ErrUserNotFound       = errors.New("usuario no encontrado")
+	ErrInvalidCredentials = errors.New("credenciales inválidas")
```
Comment (ekimike) [inline backend/internal/register/application/registerUseCase.go:11]: You meant:
user not found
invalid credentials

File: backend/internal/register/application/registerUseCase.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+func (r *RegistrationUseCase) Register(phone, email, username,password string) (*domain.User, error) {
+	
+	existingUser, err := r.repo.GetUserByEmail(email)
+	if err != nil {
+		return nil, err
+	}
+	if existingUser != nil {
+		return nil, errors.New("el nombre de usuario ya está en uso")
```
Comment (ekimike) [inline backend/internal/register/application/registerUseCase.go:33]: You meant:

the username is already in use

File: backend/internal/register/infrastructure/userRepository.go [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+	newUser, err := domain.NewUser(phone, email, username,password) // Utilizamos el correo electrónico como identificador
+	if err != nil {
+		return nil, err
+	}
+	
+
+	query := `INSERT INTO soc_app_users ( insertion_date,phone, email, user_name, password) VALUES (NOW(),$1, $2, $3, $4)`
+
+	fmt.Println("Insertando usuario en la base de datos...")
```
Comment (ekimike) [inline backend/internal/register/infrastructure/userRepository.go:58]: Whenever is needed to track down our actions, it should be better to use a logging library.

File: backend/internal/register/infrastructure/userRepository.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+
+func (u *UserDBRepository) RegisterUser( phone, email, username,password string) (*domain.User, error) {
+
+	existingUser, err := u.GetUserByEmail(email)
+	
+	if err != nil {
+		return nil, err
+	}
+	if existingUser != nil {
+		return nil, errors.New("el correo electrónico ya está en uso")
```
Comment (ekimike) [inline backend/internal/register/infrastructure/userRepository.go:46]: you meant?:

email is already in use

File: backend/internal/register/infrastructure/userRepository.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+		return nil, err
+	}
+	
+
+	query := `INSERT INTO soc_app_users ( insertion_date,phone, email, user_name, password) VALUES (NOW(),$1, $2, $3, $4)`
+
+	fmt.Println("Insertando usuario en la base de datos...")
+	tx,er:=u.db.Begin()
+	if er!=nil{
+		fmt.Println("Error al iniciar la transaccion")
```
Comment (ekimike) [inline backend/internal/register/infrastructure/userRepository.go:61]: You meant:

Error while starting the transaction

**Review by ekimike — CHANGES_REQUESTED** [review]

**Comment by AngelGio** [comment]
@miguel1808 The HTTP status code in the TestRegisterUserHandler_RegisterUser_Success test case should be 201 Created instead of 200 OK. The HTTP status code 201 Created is more appropriate for indicating that a resource has been successfully created, which aligns with the registration of a new user. Please update the test to use http.StatusCreated for a more accurate representation of the expected HTTP status code.

**Review by LuiggiPasacheL — APPROVED** [review]

**Review by ekimike — APPROVED** [review]

**Pull request merged**